### PR TITLE
Make client_secret optional for password grant type

### PR DESCRIPTION
**Pull request by chainlink**
Add tests and flag
fix up test name (clientId > clientSecret)

### CODE REVIEW

**Comment by chainlink** [comment]
Completion of https://github.com/oauthjs/node-oauth2-server/pull/312

**Comment by chainlink** [comment]
Updated!

File: lib/handlers/token-handler.js [hunk 1]
```diff
@@ -269,6 +278,17 @@ TokenHandler.prototype.updateErrorResponse = function(response, error) {
 };
 
 /**
+ * Given a grant type, check if client authentication is required
+ */
+TokenHandler.prototype.checkClientAuthentication = function(grantType) {
```
Comment (mjsalinger) [inline lib/handlers/token-handler.js]: Maybe rename this to `isClientAuthenticationRequired`

File: lib/server.js [hunk 2]
```diff
@@ -63,7 +63,10 @@ OAuth2Server.prototype.token = function(request, response, options, callback) {
   options = _.assign({
     accessTokenLifetime: 60 * 60,             // 1 hour.
     refreshTokenLifetime: 60 * 60 * 24 * 14,  // 2 weeks.
-    allowExtendedTokenAttributes: false
+    allowExtendedTokenAttributes: false,
+    requiresClientAuthentication: {
+      blah: true
```
Comment (mjsalinger) [inline lib/server.js]: What's the purpose of this key?  Why not just an empty object?

**Comment by chainlink** [comment]
Updated and rebased!

**Comment by mjsalinger** [comment]
LGTM

**Comment by thoptersword** [comment]
@mjsalinger this is incredibly useful, can we get this rolled out as 3.0.0-b3.2 soon?

**Pull request merged**